web-app/CKS.Web.Test/UnitTests/Controllers/AutocompleteControllerTests.cs

```diff
@@ -0,0 +1,198 @@
+		{
+			var searchUrl = new SearchUrl() { q = "dia" };
+			var result = autocompleteController.Get(searchUrl);
+
+			mockSearchProvider.Verify(mock => mock.TypeAhead(searchUrl), Times.Once());
+		}
+
+		[Fact]
+		public void AutocompleteReturnsEmptyListWith500ServerErrorWhenSearchFails()
+		{
+			mockSearchProvider
+				.Setup(sp => sp.TypeAhead(It.IsAny<ISearchUrl>()))
+				.Returns(new TypeAheadResults() { Failed = true });
+
+			var response = autocompleteController.Get(new SearchUrl());
+
+			response.StatusCode.ShouldBe(500);
+			((IEnumerable<object>) response.Value).Count().ShouldBe(0);
+		}
+
+		[Fact]
+		public void AutocompleteLogsErrorMessageAndSearchUrlForFailedSearch()
+		{
+			mockSearchProvider
+				.Setup(sp => sp.TypeAhead(It.IsAny<ISearchUrl>()))
+				.Returns(new TypeAheadResults() { Failed = true, ErrorMessage = "An error message" });
+
+			var response = autocompleteController.Get(new SearchUrl() { q = "can" });
+
+			mockLogger.VerifyLog(logger => logger.LogError("Autocomplete exception - An error message. Search url - search?q=can."));
+		}
+
+		[Fact]
+		public void AutocompleteReturnsResponseWithTypeAheadResultsFromSearch()
+		{
+			var responseData = (autocompleteController.Get(new SearchUrl()).Value as IEnumerable<AutocompleteItemResponse>).ToList();
+			responseData.Count().ShouldBe(2);
+		}
+
+		[Fact]
+		public void AutocompleteAppendsTopicForTitleResults()
+		{
+			var responseData = (autocompleteController.Get(new SearchUrl()).Value as IEnumerable<AutocompleteItemResponse>).ToList();
+			responseData[0].Title.ShouldBe("Diabetes - type 1 (Topic)");
+		}
+
+		[Fact]
+		public void AutocompleteUsesTypeAheadLinkForTitleResults()
+		{
+			var responseData = (autocompleteController.Get(new SearchUrl()).Value as IEnumerable<AutocompleteItemResponse>).ToList();
+			responseData[0].Link.ShouldBe("/topics/diabetes-type-1/");
+		}
+
+		[Fact]
+		public void AutocompleteUsesSERPWithTrailingSlashForKeywordResults()
+		{
+			var responseData = (autocompleteController.Get(new SearchUrl()).Value as IEnumerable<AutocompleteItemResponse>).ToList();
+			responseData[1].Link.ShouldBe("/search/?q=diabetes");
+		}
+
+		#endregion
+
+		#region Opensearch suggestions end point
+
+		[Fact]
+		public void OpenSearchCallsSearchProviderWithSearchUrl()
+		{
+			var searchUrl = new SearchUrl() { q = "dia" };
+			var result = autocompleteController.GetOpenSearchSuggestions(searchUrl);
+
+			mockSearchProvider.Verify(mock => mock.TypeAhead(searchUrl), Times.Once());
+		}
+
+		[Fact]
+		public void OpenSearchReturnsEmptyResultListWith500ServerErrorWhenSearchFails()
+		{
+			mockSearchProvider
+				.Setup(sp => sp.TypeAhead(It.IsAny<ISearchUrl>()))
+				.Returns(new TypeAheadResults() { Failed = true });
+
+			var response = autocompleteController.GetOpenSearchSuggestions(new SearchUrl { q = "diab" });
+
+			response.StatusCode.ShouldBe(500);
+
+			var resultList = ((IEnumerable<object>)response.Value).ToList();
+			resultList.Count.ShouldBe(4);
```

Comment:
If it's failed with a 500 error why do have a count of 4 results?

Reply:
Ah so it's not 4 suggestion results - the result is a list, if that makes sense. The spec says return an array with 4 items in - the first is the query term and the other 3 are suggestions, descriptions, and links. When the search request fails it makes sense to still return the data in the correct format, albeit with empty arrays. There's a comment in the class itself explaining the weird open search suggestions format.

Reply:
Ah I see OK!